File: python/binary_search.py
"""
Always remember in order for a binary search algorithm to complete the tasks the items in the list need to be sorted.
"""

def binary_search(list, target):
  first = 0
  last = len(list) - 1
  
  while first <= last:
    midpoint = (first + last)//2
    if list[midpoint] == target:
      return midpoint
    elif list[midpoint] < target:
      first = midpoint+1
    else:
      last = midpoint-1
  return -1




def verify(index):
  if index is not -1:
    print("Target found at index: ", index)
  else:
    print("Target not found in list")
    
numbers = [1,2,3,4,5,6,7,8,9,10]

verify(binary_search(numbers, 4))
##verify(result)

##result = binary_search(numbers,6)
##verify(result)

File: javascript/binary_search.js
// Binary function that eliminates possibilities that don't match specific criteria. 
	//Pre-conditions: the list must be sorted in asc order. 
	// O(log n)

function binary_search( list, target ) {
	var first = 0;
	var last = ( list.length ) -1;
	var operations = 0; // initialize operation counter.
	
	while (first <= last) {
		var mid = Math.floor( ( first + last ) / 2 );
		operations++; // add 1 if the 
		// When the 
		if( list[ mid ] == target ) {
			//return a json object that has the operations performed to get to the index as well as the index targeted
			return {'operations' : operations, 'index' : mid};
		} else if( mid < target) {
			first = mid + 1;
		} else if( mid > target ) {
			last = mid - 1;
		}
	}
	return -1;
}

function verify( json ) {
	if( json !== -1 ) {
		return console.log( 'It takes ' + json.operations + ' operations to find the index:' + json.index );
		//return console.log('Target found at index: ' + index);
	} else {
		return console.log( 'Target not found!' );
	}
}


var list = [];

for(var  i = 0; i <= 1000; i++) {
	list.push(i);
}

verify( binary_search( list, 10) );
verify( binary_search( list, 20) );
verify( binary_search( list, 30) );
verify( binary_search( list, 40) );
verify( binary_search( list, 50) );
verify( binary_search( list, 60) );
verify( binary_search( list, 80) );
verify( binary_search( list, 90) );
verify( binary_search( list, 100) );
verify( binary_search( list, 500) );